config/get_git_version.py

```diff
@@ -0,0 +1,101 @@
+# This script takes these information and generates the version.h which later is
+# used by the library to report its version.
+import argparse
+import sys
+import os
+import re
+
+
+def normalize_version(git_version, git_is_dirty):
+    if '-' in git_version:
+        cleaned = re.search('[0-9]+\.[0-9]+\.[0-9]\-[0-9]+', git_version)
+        cleaned_string = cleaned.group(0).replace("-", ".")
+    elif 'v' in git_version:
+        cleaned_string = git_version.replace("v", "")
+    else:
+        cleaned_string = git_version
+
+    # In case the repository is in a dirty state (uncommited changes)
+    # we do tell the user during build by writing to stdout.
+    # That is the way it is done in the CMake Build as well.
+    # Maybe think about adding the -dirty also for the version header.
+    if git_is_dirty == "TRUE":
+        git_version_dirty = git_version+"-dirty"
+        print("git version: " + git_version_dirty +
+              " normalized to " + cleaned_string)
+
+    return cleaned_string
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Generate version header')
+    parser.add_argument('--header',
+                        required=True,
+                        help='output header file')
+    parser.add_argument('--header_input',
+                        required=True,
+                        help='input header file')
+    parser.add_argument('--volatile_file',
+                        required=True,
+                        help='file containing the git version variables')
+    parser.add_argument('--version_variable_name',
+                        required=True,
+                        help='variablename of the hash')
+    parser.add_argument('--is_dirty_name',
+                        required=True,
+                        help='variablename of the boolean communicating if the workspace has no local changes')
+    parser.add_argument('--default_version',
+                        required=True,
+                        help='variablename for version which should be used in case git was not executable.')
+
+    args = parser.parse_args()
+
+    # Read volatile-status.txt file
+    git_version = ""
+    is_dirty = ""
+    try:
+        with open(args.volatile_file, "r") as f:
+            for entry in f.read().split("\n"):
+                if entry:
+                    key_value = entry.split(' ', 1)
+                    key = key_value[0].strip()
+                    if key == args.version_variable_name:
+                        git_version = key_value[1].strip()
+                    if key == args.is_dirty_name:
+                        is_dirty = key_value[1].strip()
+                    if key == args.default_version:
+                        default_version = key_value[1].strip()
+    except:
+        # In case volatile-status cannot be read, exit with an error
+        sys.exit("Cannot open volatile-status.txt")
+
+    if git_version == "" or is_dirty == "" or default_version == "":
```

Comment:
i think this will still have the bug. if we didn't manage to read a git_version or is_dirty then we want to return the default version.

however, if default_version is "" then we have an API contract failure and should sys.exit.

so (earlier)

```
if default_version == "":
  sys.exit("default_version must be set")
```

and here

```
if git_version == "" or is_dirty == "":
  git_version = default_version
else:
  git_version = normalize_version(git_version, is_dirty)
```

i think.

Reply:
I cant think of a case where this happens. In workspace_status.py if git cant be called or we are not in a git checkout the git_version gets set to 0.0.0 or v.0.0.0 and git is_dirty gets set to true while the default version is set correctly.
This means either all of the fields are empty or none. I just or'ed them in case anyone changes workspace_status.py.

The only case where we do not have a git_version or is_dirty is when workspace_status.py isn't called . If this can happen there is no default_version we can set to either

Reply:
@tkoeppe was the person who reported the issue to me. i think they were using a snapshot zip and then building it, so they didn't have a git checkout but were building with bazel.

did you change this behaviour in this new version?

Reply:
the build didn't start as the python script exited without outputing anything useful that bazel could use.

Reply:
Do you have at any chance a log for this. Because if the sys.exit in the python file were triggered there should be output to stderr. namely:
Cannot open volatile-status.txt or No usable entry in volatile-status.txt
If I change the if clauses so the always catch for testing purposes I get:
```
INFO: Analyzed target //:benchmark (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /Users/mad/Desktop/Verwaltung/Programs/gbench/benchmark/working2/benchmark/BUILD.bazel:21:24: Action include/benchmark/version.h failed: (Exit 1): get_git_version failed: error executing command bazel-out/host/bin/get_git_version --header bazel-out/darwin-fastbuild/bin/include/benchmark/version.h --header_input include/version.h.in --volatile_file bazel-out/volatile-status.txt ... (remaining 6 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox
No usable entry in volatile-status.txt
Target //:benchmark failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.015s, Critical Path: 0.54s
INFO: 2 processes: 2 internal.
FAILED: Build did NOT complete successfully
```

Reply:
if you encountered that I probably know how to fix that. Otherwise I don't know what happened there

Reply:
> Do you have at any chance a log for this. Because if the sys.exit in the python file were triggered there should be output to stderr. namely: Cannot open volatile-status.txt or No usable entry in volatile-status.txt If I change the if clauses so the always catch for testing purposes I get:
> 
> ```
> INFO: Analyzed target //:benchmark (0 packages loaded, 0 targets configured).
> INFO: Found 1 target...
> ERROR: /Users/mad/Desktop/Verwaltung/Programs/gbench/benchmark/working2/benchmark/BUILD.bazel:21:24: Action include/benchmark/version.h failed: (Exit 1): get_git_version failed: error executing command bazel-out/host/bin/get_git_version --header bazel-out/darwin-fastbuild/bin/include/benchmark/version.h --header_input include/version.h.in --volatile_file bazel-out/volatile-status.txt ... (remaining 6 arguments skipped)
> 
> Use --sandbox_debug to see verbose messages from the sandbox
> No usable entry in volatile-status.txt
> Target //:benchmark failed to build
> Use --verbose_failures to see the command lines of failed build steps.
> INFO: Elapsed time: 1.015s, Critical Path: 0.54s
> INFO: 2 processes: 2 internal.
> FAILED: Build did NOT complete successfully
> ```

it was the "No usable entry" output.  this is why i think the issue was that we weren't in a git checkout and the previous logic would just bail if it couldn't get a git version.

Reply:
ah okay. perfect. That means there are ways of building the lib with bazel which ignore(?) the .bazelrc. That's fixable pretty easily. Give me a sec